File: day9/puzzle2.dart
import 'dart:convert';
import 'dart:io';
import 'dart:math';

/// Extrapolates the first value of the specified [List].
int extrapolate(List<int> values) {
  var sequences = <List<int>>[];
  var sequence = values;
  while (true) {
    sequences.add(sequence);
    if ((sequence.reduce(max) == 0) && (sequence.reduce(min) == 0)) {
      break;
    }
    var differences = <int>[];
    for (var i = sequence.length - 1; i > 0; i--) {
      differences.insert(0, sequence[i] - sequence[i - 1]);
    }
    sequence = differences;
  }
  int first = 0;
  for (var i = sequences.length - 1; i >= 0; i--) {
    sequence = sequences[i];
    if (i == sequences.length - 1) {
      sequence.add(first);
    } else {
      first = sequence.first - first;
      sequence.insert(0, first);
    }
  }
  return first;
}

/// Application entry-point.
void main() async {
  final lines = utf8.decoder
      .bind(File("puzzle_input.txt").openRead())
      .transform(const LineSplitter());
  int sum = 0;
  await for (final line in lines) {
    sum += extrapolate([for (final value in line.split(' ')) int.parse(value)]);
  }
  print('sum = $sum');
}

File: day9/puzzle2.py
def extrapolate(values: list[int]) -> int:
    """Extrapolates the previous value.

    Parameters
    ----------
    values : list[int]
        The values.

    Returns
    -------
    int
        The extrapolated value.
    """
    sequences: list[list[int]] = []
    sequence = values
    while True:
        sequences.append(sequence)
        if min(sequence) == 0 and max(sequence) == 0:
            break
        differences = []
        for i in range(len(sequence) - 1, 0, -1):
            difference = sequence[i] - sequence[i - 1]
            differences.insert(0, difference)
        sequence = differences
    first = 0
    for i in range(len(sequences) - 1, -1, -1):
        sequence = sequences[i]
        if i == len(sequences) - 1:
            sequence.insert(0, first)
        else:
            first = sequence[0] - first
            sequence.insert(0, first)
    return first


def main():
    """Application entry-point."""
    sum = 0
    with open("puzzle_input.txt", "rt") as f:
        for line in f:
            values = [int(value) for value in line.split(" ")]
            sum += extrapolate(values)
    print(f"sum = {sum}")


if __name__ == "__main__":
    main()
